managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/PendingReadsManager.java

```diff
@@ -437,6 +458,60 @@ public void readEntriesFailed(ManagedLedgerException exception, Object ctx) {
         }
     }
 
+    private AsyncCallbacks.ReadEntriesCallback handlePendingReadsLimits(ReadHandle lh,
+                                                                long firstEntry, long lastEntry,
+                                                                boolean shouldCacheEntry,
+                                                                AsyncCallbacks.ReadEntriesCallback originalCallback,
+                                                                Object ctx, PendingReadsLimiter.Handle handle) {
+        if (pendingReadsLimiter.isDisabled()) {
+            return originalCallback;
+        }
+        long estimatedReadSize = (1 + lastEntry - firstEntry)
+                * (estimatedEntrySize + BOOKKEEPER_READ_OVERHEAD_PER_ENTRY);
+        final AsyncCallbacks.ReadEntriesCallback callback;
+        PendingReadsLimiter.Handle newHandle = pendingReadsLimiter.acquire(estimatedReadSize, handle);
+        if (!newHandle.success) {
+            long now = System.currentTimeMillis();
+            if (now - newHandle.creationTime > readEntryTimeoutMillis) {
+                String message = "Time-out elapsed reading from ledger "
+                        + lh.getId()
+                        + ", " + rangeEntryCache.getName()
+                        + ", estimated read size " + estimatedReadSize + " bytes"
+                        + " for " + (1 + lastEntry - firstEntry) + " entries";
+                log.error(message);
+                pendingReadsLimiter.release(newHandle);
+                originalCallback.readEntriesFailed(
+                        new ManagedLedgerException.TooManyRequestsException(message), ctx);
+                return null;
+            }
+            this.rangeEntryCache.ml.getExecutor().submitOrdered(lh.getId(), () -> {
+                readEntriesInternal(lh, firstEntry, lastEntry, shouldCacheEntry,
+                        originalCallback, ctx, newHandle);
+                return null;
+            });
+            return null;
```

Comment:
> Can we make this reactive by queuing cursors that have requested more entries and then feed those cursors as memory gets freed? 


I have thought about creating some kind of "async rate limiter", but after all you would have to keep a list (or priority queue) with the reads that are waiting for the needed amount of permits to be available.

Then the problem would be to implement some "fair" algorithm that:
- does not let the pending reads starve
- tried to be "fair" and serve the pending reads in FIFO order.


With the current mechanism we are scheduling the read on the Thread that is pinned to the ManagerLedger, and this naturally adds some kind of back pressure depending on the demands of reads of the ML.

In one broker all the ML are competing on the available memory and I don't want to keep a single global list of pending reads, as it will be really hard to make it "fair".

With this approach basically each ML will have its own list of waiting pending reads (the internal queue of the pinned executor).

Reply:
To clarify, your concern about fairness is not to have a global FIFO, but is to ensure that the QoS for independent subscriptions is fair, right? A global queue would be fair in that it would be FIFO, but could risk becoming a performance bottleneck. However, we'd only ever queue pending reads when we're over the memory threshold, so the performance is already going to be degraded.

One potential option for ensuring fairness is to put that on the rate limiting solution.

> With the current mechanism we are scheduling the read on the Thread that is pinned to the ManagerLedger, and this naturally adds some kind of back pressure depending on the demands of reads of the ML.

One potential issue for this solution is that the managed ledger thread is already responsible for a lot. This scheduled task has no back off, so it will run very frequently and could become expensive if there is contention on the PendingReadsLimiter lock.

> With this approach basically each ML will have its own list of waiting pending reads (the internal queue of the pinned executor).

It seems to me the current solution has a race where each ML is competing to get the memory permits. The race does not prevent a single ML from getting all of the permits.